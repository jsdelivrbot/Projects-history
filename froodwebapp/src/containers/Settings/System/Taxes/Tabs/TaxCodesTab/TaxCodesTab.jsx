import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  taxCodesGetRequest,
  taxCodesSaveRequest,
  taxCodesUpdateRequest,
  taxCodesDeleteRequest
} from 'redux-base/actions';
import { Row, Col } from 'react-flexbox-grid';
import { Table, Spin } from 'antd';
import { ConnectedTopFormModal, FroodSearchInput } from 'components';
import { percentageParser } from 'utils';
import { table } from 'styles/common.scss';
import fields from './modalFields';
import columns from './taxCodesHelpers';

const mapStateToProps = state => ({
  needReloadTaxCodes: state.taxes.needReloadTaxCodes,
  loadingPage: state.taxes.loadingPage,
  taxCodes: state.taxes.taxCodes,
});

const mapDispatchToProps = {
  taxCodesGetRequest,
  taxCodesSaveRequest,
  taxCodesUpdateRequest,
  taxCodesDeleteRequest
};

const newTax = {
  name: '',
  code: '',
  rate: 0
};

export class TaxCodesTab extends Component {
  constructor(props) {
    super(props);
    this.state = {
      modalVisible: false,
      modalData: newTax,
      searchValue: ''
    };
  }

  componentWillReceiveProps(nextProps) {
    // when we delete or update code
    if (nextProps.needReloadTaxCodes) {
      this.setState({
        modalVisible: false,
        modalData: undefined
      }, () => this.props.taxCodesGetRequest());
      return;
    }

    // when we create codes
    if (this.props.taxCodes.length !== nextProps.taxCodes.length) {
      this.setState({
        modalVisible: false,
        modalData: undefined
      });
    }
  }

  handleToggleModal = () => {
    this.setState({
      modalVisible: !this.state.modalVisible,
      modalData: newTax
    });
  }

  handleSave = (taxCode) => {
    if (this.state.modalData.id) {
      this.props.taxCodesUpdateRequest({
        payload: {
          id: this.state.modalData.id,
          rate: taxCode.rate
        }
      });
    } else {
      this.props.taxCodesSaveRequest({ payload: taxCode });
    }
  }

  handleActivate = (e) => {
    const codeId = e.target.id;
    const { rate } = this.props.taxCodes.find(cde => cde.id === Number(codeId));
    this.props.taxCodesUpdateRequest({
      payload: {
        id: codeId,
        rate
      }
    });
  }

  handleDeactivate = (e) => {
    this.props.taxCodesDeleteRequest({
      id: e.target.id
    });
  }

  handleEdit = (e) => {
    const codeId = e.target.id;
    const {
      id,
      name,
      code,
      rate
    } = this.props.taxCodes.find(cde => cde.id === Number(codeId));

    this.setState({
      modalVisible: true,
      modalData: {
        id,
        name,
        code,
        rate
      }
    });
  }

  handleFormatter(value) {
    return `${value}%`;
  }

  handleSearchChange = (value) => {
    this.setState({
      searchValue: value
    });
  }

  render() {
    const {
      modalData,
      modalVisible,
      searchValue
    } = this.state;

    const { taxCodes, loadingPage } = this.props;

    const data = taxCodes.filter(item =>
      item && item.name && item.name.toLowerCase().includes(searchValue.toLowerCase()));

    return (
      <div>
        <ConnectedTopFormModal
          loading={ loadingPage }
          title="New Tax Code"
          buttonText="New Tax Code"
          handleSave={ this.handleSave }
          handleToggleModal={ this.handleToggleModal }
          visible={ modalVisible }
          initialValues={ modalData }
          fields={ fields(this.handleFormatter, percentageParser) }
        />
        <Spin spinning={ loadingPage }>
          <Row>
            <Col lg={ 4 }>
              <FroodSearchInput
                onChange={ this.handleSearchChange }
                placeholder="Type to search tax codes"
              />
            </Col>
          </Row>
          <Table
            className={ table }
            rowKey="id"
            size="small"
            columns={ columns(this.handleEdit, this.handleDeactivate, this.handleActivate) }
            dataSource={ data }
          />
        </Spin>
      </div>
    );
  }
}

TaxCodesTab.propTypes = {
  // trigger
  loadingPage: PropTypes.bool.isRequired,
  needReloadTaxCodes: PropTypes.bool.isRequired,
  // data
  taxCodes: PropTypes.array.isRequired,
  // redux-base
  taxCodesGetRequest: PropTypes.func.isRequired,
  taxCodesSaveRequest: PropTypes.func.isRequired,
  taxCodesUpdateRequest: PropTypes.func.isRequired,
  taxCodesDeleteRequest: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(TaxCodesTab);
