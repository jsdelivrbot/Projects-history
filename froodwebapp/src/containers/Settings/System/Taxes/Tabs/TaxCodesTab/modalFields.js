export default (formatter, parser) => ([{
  header: 'Tax Code Name',
  type: 'text',
  name: 'name',
  editableForExistingItem: false,
  editableForNewItem: true,
}, {
  header: 'Tax Code',
  type: 'text',
  name: 'code',
  editableForExistingItem: false,
  editableForNewItem: true,
}, {
  header: 'Tax Amount',
  type: 'number',
  name: 'rate',
  min: 0,
  formatter,
  parser,
  editableForExistingItem: true,
  editableForNewItem: true,
}]);
