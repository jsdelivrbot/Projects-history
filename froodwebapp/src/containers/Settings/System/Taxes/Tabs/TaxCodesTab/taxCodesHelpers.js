import React from 'react';
import { TableActionButtons } from 'components';

export default (handleEdit, handleDeactivate, handleActivate) => [{
  title: 'Tax Name',
  dataIndex: 'name',
  width: '20%',
}, {
  title: 'Tax Code',
  dataIndex: 'code',
  width: '20%',
}, {
  title: 'Tax Rate',
  width: '20%',
  render: (text, record) => (
    `${record.rate}%`
  )
}, {
  title: 'Status',
  dataIndex: 'isActive',
  width: '20%',
  render: text => (
    text ? 'Active' : 'Not Active'
  )
}, {
  title: 'Action',
  width: '20%',
  render: (text, record) => (
    <TableActionButtons
      record={ record }
      handleEdit={ handleEdit }
      handleDeactivate={ handleDeactivate }
      handleActivate={ handleActivate }
      activateButtonVisible
    />
  )
}];
