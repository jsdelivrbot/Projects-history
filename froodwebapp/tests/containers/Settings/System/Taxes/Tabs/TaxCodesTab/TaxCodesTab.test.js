import { TaxCodesTab } from 'containers/Settings/System/Taxes/Tabs/TaxCodesTab/TaxCodesTab';
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const { describe, expect, it } = global;

describe('TaxCodesTab', () => {
  const props = {
    loadingPage: false,
    needReloadTaxCodes: false,
    taxCodes: [],
    taxCodesGetRequest: jest.fn(),
    taxCodesSaveRequest: jest.fn(),
    taxCodesUpdateRequest: jest.fn(),
    taxCodesDeleteRequest: jest.fn()
  };

  it('renders the TaxCodesTab component', () => {
    const component = shallow(<TaxCodesTab { ...props } />);

    expect(component.length).toEqual(1);
  });

  it('capturing Snapshot of TaxCodesTab', () => {
    const component = shallow(<TaxCodesTab { ...props } />);
    const view = toJson(component);
    const actual = expect(view);

    actual.toMatchSnapshot();
  });
});
