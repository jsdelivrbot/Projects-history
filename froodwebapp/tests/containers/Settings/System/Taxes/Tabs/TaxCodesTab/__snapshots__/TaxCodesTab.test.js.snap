// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaxCodesTab capturing Snapshot of TaxCodesTab 1`] = `
<div>
  <ReduxForm
    buttonText="New Tax Code"
    fields={
      Array [
        Object {
          "editableForExistingItem": false,
          "header": "Tax Code Name",
          "name": "name",
          "type": "text",
        },
        Object {
          "editableForExistingItem": false,
          "header": "Tax Code",
          "name": "code",
          "type": "text",
        },
        Object {
          "editableForExistingItem": true,
          "formatter": [Function],
          "header": "Tax Amount",
          "name": "rate",
          "parser": [Function],
          "type": "number",
        },
      ]
    }
    handleSave={[Function]}
    handleToggleModal={[Function]}
    initialValues={
      Object {
        "code": "",
        "name": "",
        "rate": 0,
      }
    }
    loading={false}
    title="New Tax Code"
    visible={false}
  />
  <Spin
    prefixCls="ant-spin"
    size="default"
    spinning={false}
    wrapperClassName=""
  >
    <Row>
      <Col
        lg={4}
      >
        <FroodSearchInput
          onChange={[Function]}
          placeholder="Type to search tax codes"
        />
      </Col>
    </Row>
    <Table
      bordered={false}
      className="table"
      columns={
        Array [
          Object {
            "dataIndex": "name",
            "title": "Tax Name",
            "width": "20%",
          },
          Object {
            "dataIndex": "code",
            "title": "Tax Code",
            "width": "20%",
          },
          Object {
            "render": [Function],
            "title": "Tax Rate",
            "width": "20%",
          },
          Object {
            "dataIndex": "isActive",
            "render": [Function],
            "title": "Status",
            "width": "20%",
          },
          Object {
            "render": [Function],
            "title": "Action",
            "width": "20%",
          },
        ]
      }
      dataSource={Array []}
      indentSize={20}
      loading={false}
      locale={Object {}}
      prefixCls="ant-table"
      rowKey="id"
      rowSelection={null}
      showHeader={true}
      size="small"
      useFixedHeader={false}
    />
  </Spin>
</div>
`;
